src/cons_tuples_impl.rs

```diff
@@ -57,8 +57,8 @@ impl<I, J> Clone for ConsTuples<I, J>
 
 /// Create an iterator that maps for example iterators of
 /// `((A, B), C)` to `(A, B, C)`.
-pub fn cons_tuples<I, J>(iterable: I) -> ConsTuples<I, J>
-    where I: Iterator<Item=J>
+pub fn cons_tuples<I, J>(iterable: I) -> ConsTuples<I::IntoIter, J>
+    where I: IntoIterator<Item=J>
```

Comment:
This seems extremely unlikely to break existing code, but I'm sure one could write code that breaks from this change. Not sure whether that's important.

Reply:
I thought this change only makes `cons_tuples` more general as per [`impl<I: Iterator> IntoIterator for I`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#impl-IntoIterator-25).

Reply:
Yeah, that's what I was thinking of too. Pretty much everything that affects public API is a breaking change for inference in some way.

Reply:
I think it doesn't.  The usual inference breaks:
- Adding a second implementation of a trait with a generic (`AsRef` and `PartialEq` are common offenders)
- Changing from a concrete type to a trait (`&` to `impl Borrow`, particularly)

In this case there's the blanket impl, so if you had enough information to pick an Iterator before, that's enough to pick an IntoIterator (because it's the same type), AFAICT.

Reply:
Yeah, this is harmless. Minimal example that compiles fine regardless of the bound on `Foo`:
```rust
mod itertools {
    pub trait IntoIterator {}
    
    pub trait Iterator {}
    
    impl<T: Iterator> IntoIterator for T {}
    
    pub fn foo<I>()
    where
        I: /*Iterator*/ IntoIterator
    {}
}

mod dependent_crate {
    pub use super::itertools::*;
    
    pub fn bar<I>()
    where
        I: Iterator
    {
        foo::<I>()
    }
}
```